ion-c-sys/src/writer.rs

```diff
@@ -524,18 +290,308 @@ impl IonCValueWriter for IonCWriterHandle<'_> {
     /// # Ok(())
     /// # }
     /// ```
+    fn start_container(&mut self, tid: ION_TYPE) -> IonCResult<()>;
+
+    /// Finishes a container.
+    fn finish_container(&mut self) -> IonCResult<()>;
+}
+
+/// The API for writing values with annotations and/or field names.
+pub trait IonCAnnotationsFieldWriter {
+    /// The associated type of value writer when writing within annotations/field context
+    type AFValueWriter: IonCValueWriter;
+
+    /// Writes a value within a context of annotations and/or a field name
+    ///
+    /// Note that it is undefined behavior if a value writing method is **not** called
+    /// or if a value writing method is called **more than once**.  For this reason,
+    /// most users should prefer the `field()` and/or `annotations()` method on the
+    /// [`IonCWriter`](./trait.IonCWriter.html) trait as it doesn't have this edge case.
+    fn write_annotations_and_field<'a, A, F, FN>(
```

Comment:
Should `write_annotations_and_field` be marked as `unsafe` given that it's on the caller to understand these requirements to avoid UB?

Reply:
It isn't `unsafe` from the memory model perspective,  just undefined if the user uses the APIs in a weird way.  In other words, there is no leak if someone were to call this method with an empty closure, it is just probably going to lead to a failure because we set up the writing of the field/annotations.

I could've made this API more robust (but chose not to), this code won't leak, but may do weird things if used improperly.  I've added #97 in case we care to make this enhancement.

Reply:
> It isn't unsafe from the memory model perspective, just undefined if the user uses the APIs in a weird way.

I'm thinking of "undefined behavior" in the C compiler sense. If it's possible to trigger UB via this API, it should probably be marked `unsafe`. However, it sounds like the undefined behavior being mentioned here would just result in a logic bug, in which case I'm happy to track changing the contract as an issue.